File: ares-fabric-1.16/src/main/java/dev/tigr/ares/fabric/utils/Comparators.java
package dev.tigr.ares.fabric.utils;

import dev.tigr.ares.Wrapper;
import dev.tigr.ares.core.feature.module.Module;
import net.minecraft.entity.Entity;
import net.minecraft.util.math.BlockPos;

import java.util.Comparator;

/**
 * @author Tigermouthbear 5/3/20
 */
public class Comparators {
    public static final EntityDistance entityDistance = new EntityDistance();
    public static final BlockDistance blockDistance = new BlockDistance();
    public static final ModuleNameLength moduleStrLength = new ModuleNameLength();
    public static final ModuleAlphabetic moduleAlphabetic = new ModuleAlphabetic();

    private static class EntityDistance implements Comparator<Entity>, Wrapper {
        @Override
        public int compare(Entity p1, Entity p2) {
            final double one = Math.sqrt(MC.player.distanceTo(p1));
            final double two = Math.sqrt(MC.player.distanceTo(p2));
            return Double.compare(one, two);
        }
    }

    private static class BlockDistance implements Comparator<BlockPos>, Wrapper {
        @Override
        public int compare(BlockPos pos1, BlockPos pos2) {
            final double one = Math.sqrt(MC.player.squaredDistanceTo(pos1.getX() + 0.5, pos1.getY() + 0.5, pos1.getZ() + 0.5));
            final double two = Math.sqrt(MC.player.squaredDistanceTo(pos2.getX() + 0.5, pos2.getY() + 0.5, pos2.getZ() + 0.5));
            return Double.compare(one, two);
        }
    }

	private static class ModuleNameLength implements Comparator<Module>, Wrapper {
        @Override
		public int compare(Module h1, Module h2) {
			final double h1Width = FONT_RENDERER.getStringWidth(h1.getHudName());
			final double h2Width = FONT_RENDERER.getStringWidth(h2.getHudName());
			return Double.compare(h2Width, h1Width);
		}
	}

    private static class ModuleAlphabetic implements Comparator<Module> {
        @Override
        public int compare(Module h1, Module h2) {
            return h1.getName().compareTo(h2.getName());
        }
    }
}

File: src/main/kotlin/org/aresclient/ares/impl/util/Comparators.kt
package org.aresclient.ares.impl.util

import net.minecraft.entity.Entity
import net.minecraft.util.math.BlockPos
import org.aresclient.ares.api.Wrapper
import kotlin.math.sqrt

object Comparators: Wrapper {
    object EntityDistance: Comparator<Entity> {
        override fun compare(p1: Entity?, p2: Entity?): Int {
            val one = sqrt(SELF.distanceTo(p1)).toDouble()
            val two = sqrt(SELF.distanceTo(p2)).toDouble()
            return one.compareTo(two)
        }
    }

    object BlockDistance: Comparator<BlockPos> {
        override fun compare(pos1: BlockPos, pos2: BlockPos): Int {
            val one = sqrt(SELF.squaredDistanceTo(pos1.x + 0.5, pos1.y + 0.5, pos1.z + 0.5))
            val two = sqrt(SELF.squaredDistanceTo(pos2.x + 0.5, pos2.y + 0.5, pos2.z + 0.5))
            return two.compareTo(one)
        }
    }
}